bin/eval_inspec_metrics.py

```diff
@@ -0,0 +1,147 @@
+from sklearn.model_selection import train_test_split
+import pandas as pd
+from dotenv import load_dotenv
+from tqdm import tqdm
+import os
+
+
+import tml_decoder.utils.common_utils as common_utils
+
+load_dotenv()
+
+run = {}
+
+if os.getenv("NEPTUNE_PROJECT") and os.getenv("NEPTUNE_API"):
+    run = neptune.init_run(
+        project=os.getenv("NEPTUNE_PROJECT"),
+        api_token=os.getenv("NEPTUNE_API_TOKEN"),
+    )
+else:
+    run = None
+
+
+class ParsedDataset(TypedDict):
+    train: Tuple[pd.DataFrame, pd.DataFrame]
+    test: Tuple[pd.DataFrame, pd.DataFrame]
+    eval: Tuple[pd.DataFrame, pd.DataFrame]
+
+
+def eval_model(
+    model: AbstractLabelModel, encoder: AbstractEncoder, parsed_dataset: ParsedDataset
+):
+    result = {"train": {}, "test": {}, "eval": {}}
+    for split_name, split in parsed_dataset.items():
+        count_cos_sim_for_ground_truth = []
+        count_cos_sim_for_avg_emb = []
+
+        X = split[0]
+        y = split[1]
+
+        texts_for_summaries = defaultdict(list)
+
+        # Iterate over the dataset and populate the defaultdict
+        for text, summaries in zip(X, y):
+            for summary in summaries:
+                texts_for_summaries[summary].append(text)
+
+        for summary, texts in tqdm(texts_for_summaries.items(), f"{split_name} split progress"):
+            generated_label = model.get_label(texts)
+            avg_embedding = encoder.average_embedding_for_texts(texts)
+
+            true_label_embedding = encoder.encode(summary)
+            generated_label_embedding = encoder.encode(generated_label)
+            cos_sim_for_ground_truth = encoder.similarity(
+                true_label_embedding, generated_label_embedding
+            )
+            cos_sim_for_avg_emb = encoder.similarity(
+                avg_embedding, generated_label_embedding
+            )
+            count_cos_sim_for_ground_truth.append(cos_sim_for_ground_truth)
+            count_cos_sim_for_avg_emb.append(cos_sim_for_avg_emb)
+
+            if run is None:
+                continue
+            
+            run[f"{split_name}/generated_label"].append(
+                {
+                    "category": summary,
+                    "generated_label": generated_label,
+                    "cos_sim_for_ground_truth": cos_sim_for_ground_truth,
+                    "cos_sim_for_avg_emb": cos_sim_for_avg_emb,
+                }
+            )
+
+        assert len(count_cos_sim_for_ground_truth) > 0, f"Length of {split_name} is 0"
+        average_cos_sim_for_gt = sum(count_cos_sim_for_ground_truth) / len(
+            count_cos_sim_for_ground_truth
+        )
+        average_cos_sim_for_avg_emb = sum(count_cos_sim_for_avg_emb) / len(
+            count_cos_sim_for_avg_emb
+        )
+        result[split_name]["avg_cos_sim_for_ground_truth"] = average_cos_sim_for_gt
+        result[split_name]["avg_cos_sim_for_avg_emb"] = average_cos_sim_for_avg_emb
+
+        if run is None:
+            continue
```

Comment:
I would strongly suggest assuming that everybody who runs this code uses Neptune. What do you think?

Reply:
:ok_hand: